framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -103,64 +114,51 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetPrototypeService(const Bund
 
 std::shared_ptr<void> ServiceReferenceBasePrivate::GetService(BundlePrivate* bundle)
 {
-  auto s = ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
-  if (!s)
-  {
-    registration->Lock(), --registration->dependents[bundle];
-  }
-  return s;
+  return ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
 }
 
 InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundlePrivate* bundle)
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
   {
     auto l = registration->Lock(); US_UNUSED(l);
     if (!registration->available) return s;
     serviceFactory = std::static_pointer_cast<ServiceFactory>(
           registration->GetService_unlocked("org.cppmicroservices.factory"));
-    count = registration->dependents[bundle];
-  }
 
-  if (!serviceFactory)
-  {
-    auto l = registration->Lock(); US_UNUSED(l);
-          s = registration->service;
-    if (s && !s->empty()) ++registration->dependents[bundle];
-  }
-  else
-  {
-    if (count == 0)
-    {
-      s = GetServiceFromFactory(bundle, serviceFactory);
+    registration->dependents.insert(std::make_pair(bundle, 0));
 
-      auto l = registration->Lock(); US_UNUSED(l);
-      if (registration->dependents[bundle] == 0)
-      {
-        registration->bundleServiceInstance.insert(std::make_pair(bundle, s));
-      }
-      else
+    if (!serviceFactory)
+    {
+      s = registration->service;
+      if (s && !s->empty())
       {
-        // There was a race and we now have one instance too much. Return the
-        // already produced instance and ignore the additional one.
-        s = registration->bundleServiceInstance[bundle];
+        ++registration->dependents.at(bundle);
       }
-
-      if (s && !s->empty()) ++registration->dependents[bundle];
-    }
-    else
-    {
-      auto l = registration->Lock(); US_UNUSED(l);
-      // return the already produced instance
-      s = registration->bundleServiceInstance[bundle];
-      if (s && !s->empty()) ++registration->dependents[bundle];
+      return s;
     }
   }
+
+  // Calling into a service factory could cause re-entrancy into the
+  // framework and even, theoretically, into this function. Ensuring
+  // we don't hold a lock while calling into the service factory eliminates
+  // the possibility of a deadlock. It does not however eliminate the
+  // possbilituy of infinite recursion.
+  s = GetServiceFromFactory(bundle, serviceFactory);
```

Comment:
At this point, we are unlocked and know, that we have a service factory. Previously, we would try to reuse a previously cached service instance but we were missing a check for the race condition. I think we can still avoid calling `GetServiceFromFactory` every time and reuse the cached version while checking a possible race:

Case 1 `count < 1`: This case was handled fine in the original code. We call `GetServiceFromFactory`, then lock and check the count again. If it is still < 1 we insert the new instance into our cache (the map) and are done. If It is >= 1 now, we discard the newly created instance and return the (recently added) instance from the cache.

Case 2 `count >= 1`: Lock and get the cached instance. Now (this was missing originally) if the instance is `null` (because of the race) we need to go back to case 1 (probably do case 1 and 2 in a `while(!s)` loop. If it is not `null`, we increase the count and return.

Did I miss something?